main/src/main/scala/sbt/internal/TaskProgress.scala

```diff
@@ -9,99 +9,100 @@ package sbt
-          case Some(t) if t != this => currentProgressThread.set(Some(t))
-          case _                    =>
-        }
-      }
-    @tailrec override def run(): Unit = {
-      if (!isClosed.get() && (!hasReported.get || active.nonEmpty)) {
-        try {
-          if (activeExceedingThreshold.nonEmpty) doReport()
-          val duration =
-            if (firstTime.compareAndSet(true, activeExceedingThreshold.isEmpty)) threshold
-            else sleepDuration
-          val limit = duration.fromNow
-          while (Deadline.now < limit && !isClosed.get && active.nonEmpty) {
-            var task = tasks.poll((limit - Deadline.now).toMillis, TimeUnit.MILLISECONDS)
-            while (task != null) {
-              if (containsSkipTasks(Vector(task)) || lastTaskCount.get == 0) doReport()
-              task = tasks.poll
-              tasks.clear()
-            }
-          }
-        } catch {
-          case _: InterruptedException =>
-            isClosed.set(true)
-            // One last report after close in case the last one hadn't gone through yet.
-            doReport()
-
-        }
-        run()
-      } else {
-        resetThread()
+  private[this] val reportLoop = new AtomicReference[AutoCloseable]
+  private[this] val active = new ConcurrentHashMap[Task[_], AutoCloseable]
+  private[this] val nextReport = new AtomicReference(Deadline.now)
+  private[this] val scheduler =
+    Executors.newSingleThreadScheduledExecutor(r => new Thread(r, "sbt-progress-report-scheduler"))
+  private[this] val pending = new java.util.Vector[java.util.concurrent.Future[_]]
+  private def schedule[R](duration: FiniteDuration, recurring: Boolean)(f: => R): AutoCloseable = {
+    val cancelled = new AtomicBoolean(false)
+    val runnable: Runnable = () => {
+      if (!cancelled.get) {
+        try Util.ignoreResult(f)
+        catch { case _: InterruptedException => }
       }
     }
-
-    def addTask(task: Task[_]): Unit = tasks.put(task)
-
-    override def close(): Unit = {
-      isClosed.set(true)
-      interrupt()
-      report()
-      appendProgress(ProgressEvent("Info", Vector(), None, None, None))
-      resetThread()
+    val delay = duration.toMillis
+    val future =
+      if (recurring) scheduler.schedule(runnable, delay, TimeUnit.MILLISECONDS)
+      else scheduler.scheduleAtFixedRate(runnable, delay, delay, TimeUnit.MILLISECONDS)
+    pending.add(future)
+    () => Util.ignoreResult(future.cancel(true))
+  }
+  private[this] val executor =
+    Executors.newSingleThreadExecutor(r => new Thread(r, "sbt-task-progress-report-thread"))
+  override def close(): Unit = {
+    Option(reportLoop.get).foreach(_.close())
+    pending.forEach(f => Util.ignoreResult(f.cancel(true)))
+    pending.clear()
+    scheduler.shutdownNow()
+    executor.shutdownNow()
+    if (!executor.awaitTermination(1, TimeUnit.SECONDS) ||
+        !scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
+      throw new TimeoutException
```

Comment:
This `TimeoutException` is getting thrown quite a lot when I try to deploy builds on Heroku, I'm wondering if there's anything that can be done to avoid that? This is with sbt 1.6.0-M1, Java 11, on the small open-source project https://github.com/guardian/repo-genesis with PR https://github.com/guardian/repo-genesis/pull/6 .

Here's [a full build log](https://gist.github.com/rtyley/30195e8bc1fcaf01597b9c134ea47f65) - of the **7** most recent deploy attempts I've made, only **1** of them has passed - the other 6 have all failed with this `TimeoutException` causing deployment to abort...

![image](https://user-images.githubusercontent.com/52038/143140794-987240a7-227a-4afb-96d2-4f1cffd55ba6.png)

I'm using the early sbt milestone [1.6.0-M1](https://github.com/sbt/sbt/releases/tag/v1.6.0-M1) for the fix in https://github.com/sbt/sbt/pull/6649 for the `ClassCastException` problems with https://github.com/sbt/sbt/issues/6592 and Java 11 - but this timeout is maybe too aggressive for Heroku?

Reply:
Interesting. Let's extend this to 30s and/or make it configurable.

Reply:
Out of curiosity, what would happen if there was no timeout at all? I don't really know what's going on, but is it _necessary_ to block `close()` to wait for the threads to complete - can something bad happen if the `scheduler` & `executor` are not terminated after `close()` has been called?

Reply:
I guess generally we don't want any threads or memory to leak (or would like to know if some leak is happening), esp when as sbt server a session could be up for days as daemon.